File: src/consensus.rs
//! WAIFU L1 - Proof-of-Intelligence Consensus

use crate::types::*;
use crate::llm_pricer::{LlmPricer, PricerError};
use crate::state::AgenticState;
use rayon::prelude::*;
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum ConsensusError {
    #[error("Pricing failed: {0}")]
    PricingError(#[from] PricerError),
    #[error("Invalid PoI proof: {0}")]
    InvalidProof(String),
    #[error("Block validation failed: {0}")]
    ValidationFailed(String),
    #[error("Non-deterministic inference")]
    NonDeterministic,
}

pub struct ProofOfIntelligence {
    pub min_stake: f64,
    pub block_time_ns: u128,
    pub max_tx_per_block: usize,
}

impl Default for ProofOfIntelligence {
    fn default() -> Self {
        Self {
            min_stake: 1000.0,
            block_time_ns: 10_000_000,
            max_tx_per_block: 100_000,
        }
    }
}

impl ProofOfIntelligence {
    pub async fn validate_and_price(
        transactions: Vec<Transaction>,
        llm: &Arc<LlmPricer>,
        state: &Arc<AgenticState>,
    ) -> Result<String, ConsensusError> {
        if !llm.is_deterministic() {
            return Err(ConsensusError::NonDeterministic);
        }

        // Parallel pricing via rayon
        let priced: Vec<PricedTransaction> = transactions
            .par_iter()
            .filter_map(|tx| {
                let (price, rationale) = llm.price_transaction(tx, state).ok()?;
                let result = Self::execute_tx(tx, price, state);
                Some(PricedTransaction {
                    original: tx.clone(),
                    clearing_rate: price,
                    pricing_rationale: rationale,
                    result,
                })
            })
            .collect();

        let llm_root = Self::compute_llm_root(&priced, llm);
        let poi = Self::create_poi(llm, &priced);

        let block = Block {
            height: state.get_block_height() + 1,
            hash: [0u8; 32],
            parents: state.get_dag_tips(),
            transactions: priced,
            llm_state_root: llm_root,
            validator: AgentId::genesis(),
            poi_proof: poi,
            timestamp: SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_nanos(),
        };

        let hash = block.compute_hash();
        let mut final_block = block;
        final_block.hash = hash;

        state.apply_block(&final_block)
            .map_err(|e| ConsensusError::ValidationFailed(e.to_string()))?;

        Ok(hex::encode(hash))
    }

    fn execute_tx(tx: &Transaction, rate: f64, state: &AgenticState) -> ExecutionResult {
        match &tx.context.operation {
            Operation::Transfer { amount } => {
                match state.transfer_compute(&tx.from, &tx.to, *amount) {
                    Ok(_) => ExecutionResult::Success { 
                        gas_used: rate, 
                        state_changes: vec![] 
                    },
                    Err(e) => ExecutionResult::Failure { 
                        reason: e.to_string(), 
                        gas_used: rate * 0.5 
                    },
                }
            }
            _ => ExecutionResult::Success { gas_used: rate, state_changes: vec![] },
        }
    }

    fn compute_llm_root(txs: &[PricedTransaction], llm: &LlmPricer) -> [u8; 32] {
        let mut h = blake3::Hasher::new();
        h.update(&llm.get_model_hash());
        for tx in txs {
            h.update(&tx.clearing_rate.to_le_bytes());
        }
        *h.finalize().as_bytes()
    }

    fn create_poi(llm: &LlmPricer, txs: &[PricedTransaction]) -> PoIProof {
        let mut ih = blake3::Hasher::new();
        let mut oh = blake3::Hasher::new();
        for tx in txs {
            ih.update(&tx.original.hash());
            oh.update(&tx.clearing_rate.to_le_bytes());
        }
        PoIProof {
            model_hash: llm.get_model_hash(),
            input_hash: *ih.finalize().as_bytes(),
            output_hash: *oh.finalize().as_bytes(),
            temperature: 0.0,
            signature: [0u8; 64],
        }
    }
}

pub struct BlockProducer {
    pub validator_id: AgentId,
    pending: crossbeam::queue::SegQueue<Transaction>,
}

impl BlockProducer {
    pub fn new(id: AgentId) -> Self {
        Self { validator_id: id, pending: crossbeam::queue::SegQueue::new() }
    }

    pub fn submit(&self, tx: Transaction) { self.pending.push(tx); }
    pub fn pending_count(&self) -> usize { self.pending.len() }

    pub async fn produce(
        &self, llm: &Arc<LlmPricer>, state: &Arc<AgenticState>,
    ) -> Result<String, ConsensusError> {
        let mut txs = Vec::with_capacity(100_000);
        while let Some(tx) = self.pending.pop() {
            txs.push(tx);
            if txs.len() >= 100_000 { break; }
        }
        if txs.is_empty() {
            return Err(ConsensusError::ValidationFailed("empty".into()));
        }
        ProofOfIntelligence::validate_and_price(txs, llm, state).await
    }
}

File: waifu_layer/consensus.py
"""WAIFU L1 - Proof-of-Intelligence Consensus."""

from __future__ import annotations

import queue
import time
from dataclasses import dataclass

from .llm_pricer import LlmPricer, PricerError
from .state import AgenticState, StateError
from .types import AgentId, Block, ExecutionResult, Operation, PoIProof, PricedTransaction, Transaction


class ConsensusError(Exception):
    pass


class PricingError(ConsensusError):
    def __init__(self, cause: PricerError):
        super().__init__(f"Pricing failed: {cause}")
        self.cause = cause


class ValidationFailed(ConsensusError):
    def __init__(self, detail: str):
        super().__init__(f"Block validation failed: {detail}")


class NonDeterministic(ConsensusError):
    def __init__(self):
        super().__init__("Non-deterministic inference")


@dataclass
class ProofOfIntelligence:
    min_stake: float = 1000.0
    block_time_ns: int = 10_000_000
    max_tx_per_block: int = 100_000

    @staticmethod
    async def validate_and_price(transactions: list[Transaction], llm: LlmPricer, state: AgenticState) -> str:
        if not llm.is_deterministic():
            raise NonDeterministic()

        # Parallel pricing via a thread pool (rayon::par_iter equivalent -- see llm_pricer.price_batch)
        priced: list[PricedTransaction] = []
        for tx, outcome in zip(transactions, llm.price_batch(transactions, state)):
            if isinstance(outcome, Exception):
                continue
            price, rationale = outcome
            result = ProofOfIntelligence._execute_tx(tx, price, state)
            priced.append(PricedTransaction(original=tx, clearing_rate=price, pricing_rationale=rationale, result=result))

        llm_root = ProofOfIntelligence._compute_llm_root(priced, llm)
        poi = ProofOfIntelligence._create_poi(llm, priced)

        block = Block(
            height=state.get_block_height() + 1,
            hash=bytes(32),
            parents=state.get_dag_tips(),
            transactions=priced,
            llm_state_root=llm_root,
            validator=AgentId.genesis(),
            poi_proof=poi,
            timestamp=time.time_ns(),
        )

        block_hash = block.compute_hash()
        block.hash = block_hash

        try:
            state.apply_block(block)
        except StateError as e:
            raise ValidationFailed(str(e)) from e

        return block_hash.hex()

    @staticmethod
    def _execute_tx(tx: Transaction, rate: float, state: AgenticState) -> ExecutionResult:
        op: Operation = tx.context.operation
        if op.kind == "Transfer":
            try:
                state.transfer_compute(tx.from_, tx.to, op.amount)
                return ExecutionResult.success_result(gas_used=rate)
            except StateError as e:
                return ExecutionResult.failure(reason=str(e), gas_used=rate * 0.5)
        return ExecutionResult.success_result(gas_used=rate)

    @staticmethod
    def _compute_llm_root(txs: list[PricedTransaction], llm: LlmPricer) -> bytes:
        import blake3

        h = blake3.blake3()
        h.update(llm.get_model_hash())
        for tx in txs:
            h.update(_f64_le(tx.clearing_rate))
        return h.digest()

    @staticmethod
    def _create_poi(llm: LlmPricer, txs: list[PricedTransaction]) -> PoIProof:
        import blake3

        ih = blake3.blake3()
        oh = blake3.blake3()
        for tx in txs:
            ih.update(tx.original.hash())
            oh.update(_f64_le(tx.clearing_rate))
        return PoIProof(
            model_hash=llm.get_model_hash(),
            input_hash=ih.digest(),
            output_hash=oh.digest(),
            temperature=0.0,
            signature=bytes(64),
        )


class BlockProducer:
    def __init__(self, validator_id: AgentId):
        self.validator_id = validator_id
        self._pending: "queue.Queue[Transaction]" = queue.Queue()

    def submit(self, tx: Transaction) -> None:
        self._pending.put(tx)

    def pending_count(self) -> int:
        return self._pending.qsize()

    async def produce(self, llm: LlmPricer, state: AgenticState) -> str:
        txs: list[Transaction] = []
        while len(txs) < 100_000:
            try:
                txs.append(self._pending.get_nowait())
            except queue.Empty:
                break
        if not txs:
            raise ValidationFailed("empty")
        return await ProofOfIntelligence.validate_and_price(txs, llm, state)


def _f64_le(value: float) -> bytes:
    import struct

    return struct.pack("<d", value)
